revisions to command line git (#164)
merely-useful/merely-useful.github.io: rse-advanced-git.Rmd

    your starting point is up to date
    (i.e., that the changes you committed while working on the cluster yesterday are now in your local copy,
    or that the fix your colleague did last week has been incorporated).

3.  `git checkout -b name-of-feature` to create a new branch.
    The branch's name should be as descriptive as the name of a function
    or of a file containing source code:
    `"stuff"` or `"fixes"` saves you a couple of seconds of typing when you create the branch,
    and will cost you much more time than that later on
    when you're juggling half a dozen branches
    and trying to figure out what's in each.

4.  Do the work you want to do.
    If some other task occurs to you while you're doing it---for example,
    if you're writing a new function and realize that the documentation for some other function should be updated---*don't*
    do that work in this branch just because you happen to be there.
    Instead,
    commit your changes,
    switch back to `master`,
    create a new branch for the other work,
    and carry on.

5.  Periodically,
    you may `git push origin name-of-feature` to save your work to GitHub.

> 6.  When the work is done,

Reviewer: maybe worth being more explicit about the commands needed here -- that you need to be in master, pull, *then* switch to feature and merge? it's opposite of the example from before.

    you `git pull origin master` to get any changes that anyone else has put in the master copy
    into your branch
    and merge any conflicts.
    This is an important step:
    you want to do the merge *and test that everything still works* in *your* branch,
    not in `master`.

7.  Finally,
    use `git checkout master` to switch back to the `master` branch on your machine
    and `git merge name-of-feature master` to merge your changes into `master`.
    Test it one more time,
    and then `git push origin master` to send your changes to the remote repository
    for everyone else to admire and use.

> ### What *is* a feature?

Reviewer: does this need a link/anchor?

Author: Looks like in most other lessons level two headings have an anchor but not level three.

"Someone would like to merge branch A into branch X".
The pull request does *not* contain the changes:
instead,
it points at two branches,
so that if either branch is changed,
the differences displayed are always up to date.

FIXME: figure

Pull requests can be created between two branches in a single repository,
or between branches of different repositories.
The latter is common in projects with many contributors:
each contributor works in a branch in their own [fork](glossary.html#fork) of the repository,
and when they're done,
they create a pull request offering to merge their changes into the `master` branch
of the original repository.
In order to update their own repository,
they pull changes from the original:

FIXME: figure

A pull request can store more than the location of the source and destination branch.
In particular,
it can store comments that people have made about the proposed merge.
GitHub and other [forges](glossary.html#forge) allow users to add comments on the pull request as a whole,
or on particular lines,
and can mark old comments as out of date if the lines of code the comment is attached to are updated.

> 

Reviewer: ```suggestion
### Setting up contribution repositories
```

*TODO: Figure - Clone Origin [../fig/github_screenshot_cloneOrigin.png](https://github.com/UW-Madison-DataScience/git-novice-custom/tree/gh-pages/fig/github_screenshot_cloneOrigin.png)*

In terminal or Gitbash, navigate to a folder you'd like to hold this repo,
we will place it on our `Desktop`.
Once there you can use the `clone` command with the link you copied as the first argument.

```shell
$ cd ~/Desktop
$ git clone https://github.com/USERNAME/countries.git
```

> **Why does the command above say 'USERNAME'?**
>
> So that we can't copy the command above and accidentally clone someone else's
> version of countries to our computer, the command above uses the placeholder
> 'USERNAME' where you should put your own username if your copied from above
> instead of copying the link from your browser and pasting it into the command.

```text
Cloning into 'countries'...
remote: Counting objects: 6, done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 6 (delta 0), reused 6 (delta 0), pack-reused 0
Unpacking objects: 100% (6/6), done.
```

Next we will set up a connection or **remote** to the authoritative repository 

> (the original version given to you by your instructor).

Reviewer: ```suggestion
(the original Github repository given to you by your instructor).
```

Then back in your terminal, navigate into the cloned repo and add the remote 
connection to this repository.
For this command we must give the remote a different nickname, 
where our original remote is 'origin'
this new remote will be called 'upstream'.
You could give it a different nickname but 'upstream' is a common nickname for
the authoritative repository.

```shell
$ cd countries
$ git remote add upstream https://github.com/INSTRUCTOR-GIVEN/countries.git
```

> **If you tried copying the command above...**
>
> You will have to replace 'INSTRUCTOR-GIVEN' with the site your instructor
> indicated at the beginning of this lesson. This will vary depending
> on how your instructor set up for this lesson.

At anytime you can see the remote connections your repo has using the following command:

```shell
$ git remote -v
```

```text
origin	https://github.com/USERNAME/countries.git (fetch)
origin	https://github.com/USERNAME/countries.git (push)
upstream	https://github.com/INSTRUCTOR-GIVEN/countries.git (fetch)
upstream	https://github.com/INSTRUCTOR-GIVEN/countries.git (push)
```

> 

Reviewer: ```suggestion
### Submitting a pull request
```

> Now that we have this setup done we will be able to suggest 

Reviewer: Also, somewhere I would like a cheatsheet (as shown in the branches section!) of the steps you take to set up a new PR: 
1. change to master, pull changes from upstream
2. update fork with master
3. create branch, commit changes
4. push branch
5. submit PR
6. make more changes, commit, push to update PR

Or something like that. Maybe at the end of this section, before conflicts?

Author: Good idea. I've added this to the very end of the pull request section.

$ cat france.txt
```

```text
Population: 66,991,000
Capital: Paris
Largest City: Paris
```

Next we will add and commit these changes.
Then we can push them to our fork of the repo.

```shell
$ git add france.txt
$ git commit -m "Added largest city to france file"
$ git push origin addFrance
```

```text
Counting objects: 3, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (3/3), 387 bytes | 0 bytes/s, done.
Total 3 (delta 0), reused 0 (delta 0)
To https://github.com/USERNAME/countries.git
   31aa2e3..609acfe  addFrance -> addFrance
```

If we reload the pull request, we'll see that the new commit was added to the 
pull request and the changes have been automatically updated.
New commits pushed to the same branch are included in the previous pull request.
If you want to suggest changes separately you need to make separate branches but 
if you want the changes to be considered together you should put them in the same branch.

*TODO: Figure - New Commit [../fig/github_screenshot_after_new_commit.png](https://github.com/UW-Madison-DataScience/git-novice-custom/tree/gh-pages/fig/github_screenshot_after_new_commit.png)*

> 

Reviewer: ```suggestion
### Dealing with Conflicting Changes
```

```

```text
[addFrance 66d7ebf] Added France to list of countries in README
 1 file changed, 1 insertions(+), 1 deletions(-)
```

```shell
$ git push origin addFrance
```

```text
Counting objects: 3, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (3/3), 376 bytes | 0 bytes/s, done.
Total 3 (delta 1), reused 0 (delta 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/USERNAME/countries.git
   609acfe..66d7ebf  addFrance -> addFrance
```

Now if we reload the page we had a pull request we notice that our `addFrance`
branch is conflicting with upstream's `master` branch.
This is because someone else edited the same line of the `README.md` file by 
adding 'United States' where we added 'France'.

*TODO: Figure - Conflicting PR [../fig/github_screenshot_conflicting_PR.png](https://github.com/UW-Madison-DataScience/git-novice-custom/tree/gh-pages/fig/github_screenshot_conflicting_PR.png)*

In this case, it is possible to resolve this conflict in github by 
clicking the 'Resolve Conflicts' button.
However, we will reuse the skills we learned earlier to resolve this conflict locally,

> as we did in our branching conflict.

Reviewer: Did we have a branching conflict...?

Author: Nope. Well spotted!
